### PR TITLE
Improve duration humanization.

### PR DESCRIPTION
**Pull request by balamurugana**
This patch makes update message carries highest component of duration
and skips lower components eg. for duration 3 days, 9 hours, 47
minutes and 20 seconds, humanizeDuration() returns highest component 3
days.

## How Has This Been Tested?

`make test`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/update-notifier.go [hunk 1]
```diff
@@ -27,6 +27,61 @@ import (
 	"github.com/fatih/color"
 )
 
+// humanizeDuration converts time.Duration to human readable duration.
+func humanizeDuration(duration time.Duration) (durationStr string) {
```
Comment (harshavardhana) [inline cmd/update-notifier.go]: Why not deprecate this and use - https://github.com/dustin/go-humanize/blob/master/times.go

**Review by harshavardhana — COMMENTED** [review]
https://github.com/dustin/go-humanize/blob/master/times.go - take a look at this as a replacement instead of keeping our own code @balamurugana

File: cmd/update-notifier.go (hunk 1 above)
Comment (balamurugana) [reply cmd/update-notifier.go]: Humanization is slight different here.  As this affects UI, it would also be reviewed by @abperiasamy

**Comment by harshavardhana** [comment]
> This patch makes update message carries highest component of duration
> and skips lower components eg. for duration 3 days, 9 hours, 47
> minutes and 20 seconds, humanizeDuration() returns highest component 3
> days.

Like the idea of skipping in that case better is to use - humanize package instead. 

```
This lets you take a time.Time and spit it out in relative terms. For example, 12 seconds ago or 3 days from now.

fmt.Printf("This was touched %s.", humanize.Time(someTimeInstance)) // This was touched 7 hours ago.
```

**Comment by harshavardhana** [comment]
```go
package main

import (
        "fmt"
        "time"

        humanize "github.com/dustin/go-humanize"
)

func main() {
        someTimeInstance := time.Now().Add(-300 * time.Hour)
        fmt.Printf("Minio released %s\n", humanize.Time(someTimeInstance)) 
}
```

```
go run /tmp/time.go
Minio released 1 week ago
``` 

This is more readable and nicer @balamurugana

**Comment by harshavardhana** [comment]
With custom formatting.

```go
package main

import (
        "fmt"
        "time"

        humanize "github.com/dustin/go-humanize"
)

func main() {
        someTimeInstance := time.Now().UTC().Add(-100 * time.Hour)
        magnitudes := []humanize.RelTimeMagnitude{
                {time.Second, "now", time.Second},
                {2 * time.Second, "1 second %s", 1},
                {time.Minute, "%d seconds %s", time.Second},
                {humanize.Day - time.Second, "%d minutes %s", time.Minute},
                {humanize.Day, "%d hours %s", time.Hour},
                {2 * humanize.Day, "1 day %s", 1},
                {humanize.Week, "%d days %s", humanize.Day},
                {2 * humanize.Week, "1 week %s", 1},
                {6 * humanize.Month, "%d weeks %s", humanize.Week},
                {humanize.Year, "%d months %s", humanize.Month},
        }
        customRelTime := func(then time.Time) string {
                return humanize.CustomRelTime(then, time.Now().UTC(), "old", "from now", magnitudes)
        }
        fmt.Printf("Minio is %s\n", customRelTime(someTimeInstance)) // This was touched 7 hours ago.                                             
}
```

```
go run /tmp/time.go
Minio is 4 days old
```

**Comment by balamurugana** [comment]
Currently we print `Minio is X days old` whereas `github.com/dustin/go-humanize` package prints `Minio released 1 week ago`

I would like to know which way is preferred.

**Comment by harshavardhana** [comment]
> Currently we print Minio is X days old whereas github.com/dustin/go-humanize package print Minio released 1 week ago
> 
> I would like to know which way is preferred.

We can customize it as well @balamurugana - posted that example as well.

File: cmd/update-notifier.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/update-notifier.go]: Already showed you how to address this without changing the UI and also remove custom code @balamurugana

**Comment by balamurugana** [comment]
> We can customize it as well @balamurugana - posted that example as well.

The example looks like a hack. Moreover it doesn't work as expected.  I listed some cases below
```
someTimeInstance := time.Now().UTC().Add(time.Duration(-7878)) = Minio is now
someTimeInstance := time.Now().UTC().Add(time.Duration(-7878788817276)) = Minio is 131 minutes old
someTimeInstance := time.Now().UTC().Add(time.Duration(-7878788817276557)) = Minio is 13 weeks old
someTimeInstance := time.Now().UTC().Add(time.Duration(-78787888172)) = Minio is 1 minutes old
```

**Comment by harshavardhana** [comment]
> The example looks like a hack. Moreover it doesn't work as expected. I listed some cases below

It is not a hack, this is how Go behaves see the example here - https://play.golang.org/p/_UE9pHVQe5.  

You have to use the time.Duration properly along with a proper multiplier for the precision.  So i would ask you to take a look at it properly, this change in code is not useful. @balamurugana

**Comment by balamurugana** [comment]
> It is not a hack, this is how Go behaves see the example here - https://play.golang.org/p/_UE9pHVQe5.
> 
> You have to use the time.Duration properly along with a proper multiplier for the precision. So i would ask you to take a look at it properly, this change in code is not useful. @balamurugana

I am not able to understand what do you mean by https://play.golang.org/p/_UE9pHVQe5.  The examples I provided are moving back to some time from now to represent a time in past and trying to see how the older message printed.,  In your example a good duration like `12*time.Hour` is the best case

> someTimeInstance := time.Now().UTC().Add(time.Duration(-7878)) = Minio is now

For this why it is printing `now`. the time is -7.878µs older.

> someTimeInstance := time.Now().UTC().Add(time.Duration(-7878788817276)) = Minio is 131 minutes old

Why its 131 minutes than 2 hours.

> someTimeInstance := time.Now().UTC().Add(time.Duration(-7878788817276557)) = Minio is 13 weeks old

instead of 13 weeks, 3 months is better.

> someTimeInstance := time.Now().UTC().Add(time.Duration(-78787888172)) = Minio is 1 minutes old

grammatically `1 minute` is better.

I see go-humanize package wrt time is not working as expected for us.

**Comment by harshavardhana** [comment]
> For this why it is printing now. the time is -7.878µs older.

Are you being serious? 7.87 micro sec is definitely now in a humanized sense.  I don't see what you don't understand. 

Here is another example showing how you can achieve addressing your concerns.
```go
go run /tmp/time.go 
Minio is now
Minio is 2 hours old
Minio is 3 months old
Minio is 1 minute old
```

```go
package main

import (
	"fmt"
	"math"
	"time"

	humanize "github.com/dustin/go-humanize"
)

func main() {
	var defaultMagnitudes = []humanize.RelTimeMagnitude{
		{time.Second, "now", time.Second},
		{2 * time.Second, "1 second %s", 1},
		{time.Minute, "%d seconds %s", time.Second},
		{2 * time.Minute, "1 minute %s", 1},
		{time.Hour, "%d minutes %s", time.Minute},
		{2 * time.Hour, "1 hour %s", 1},
		{humanize.Day, "%d hours %s", time.Hour},
		{2 * humanize.Day, "1 day %s", 1},
		{humanize.Week, "%d days %s", humanize.Day},
		{2 * humanize.Week, "1 week %s", 1},
		{humanize.Month, "%d weeks %s", humanize.Week},
		{2 * humanize.Month, "1 month %s", 1},
		{humanize.Year, "%d months %s", humanize.Month},
		{18 * humanize.Month, "1 year %s", 1},
		{2 * humanize.Year, "2 years %s", 1},
		{humanize.LongTime, "%d years %s", humanize.Year},
		{math.MaxInt64, "a long while %s", 1},
	}
	customRelTime := func(then time.Time) string {
		return humanize.CustomRelTime(then, time.Now().UTC(), "old", "new", defaultMagnitudes)
	}
	someTimeInstance := time.Now().UTC().Add(time.Duration(-7878))
	fmt.Printf("Minio is %s\n", customRelTime(someTimeInstance))
	someTimeInstance = time.Now().UTC().Add(time.Duration(-7878788817276))
	fmt.Printf("Minio is %s\n", customRelTime(someTimeInstance))
	someTimeInstance = time.Now().UTC().Add(time.Duration(-7878788817276557))
	fmt.Printf("Minio is %s\n", customRelTime(someTimeInstance))
	someTimeInstance = time.Now().UTC().Add(time.Duration(-78787888172))
	fmt.Printf("Minio is %s\n", customRelTime(someTimeInstance))
}
```

**Comment by harshavardhana** [comment]
Sent you a PR as well https://github.com/balamurugana/minio/pull/6  @balamurugana

**Comment by balamurugana** [comment]
> Are you being serious? 7.87 micro sec is definitely now in a humanized sense. I don't see what you don't understand.

Nope. But you missed the point of update message.  We found that there is an update available where current minio is just old.  We would need to convey this message somehow for any non-zero duration.

Is the below you want?

```go
func humanizeDuration(duration time.Duration) string {
	var updateRelTimeMagnitude = []humanize.RelTimeMagnitude{
		{time.Second, "just", time.Second},
		{2 * time.Second, "1 second", 1},
		{time.Minute, "%d seconds", time.Second},
		{2 * time.Minute, "1 minute", 1},
		{time.Hour, "%d minutes", time.Minute},
		{2 * time.Hour, "1 hour", 1},
		{humanize.Day, "%d hours", time.Hour},
		{2 * humanize.Day, "1 day", 1},
		{humanize.Week, "%d days", humanize.Day},
		{2 * humanize.Week, "1 week", 1},
		{humanize.Month, "%d weeks", humanize.Week},
		{2 * humanize.Month, "1 month", 1},
		{humanize.Year, "%d months", humanize.Month},
		{18 * humanize.Month, "1 year", 1},
		{2 * humanize.Year, "2 years", 1},
		{humanize.LongTime, "%d years", humanize.Year},
	}

	utcNow := time.Now().UTC()
	pastUTCNow := time.Now().UTC().Add(duration)
	return humanize.CustomRelTime(pastUTCNow, utcNow, "", "", updateRelTimeMagnitude)
}

func main() {
	durations := []time.Duration{
		time.Duration(-78787687),
		time.Duration(-2000000000),
		time.Duration(-10000000000),
		time.Duration(-7878788817276),
		time.Duration(-7878788817276557),
		time.Duration(-78787888172),
		time.Duration(-7878788817200000000),
		time.Duration(time.Second),
		time.Duration(time.Minute),
	}

	for _, duration := range durations {
		fmt.Println("Minio is", humanizeDuration(duration), "old")
	}
}
```
Output:
```
Minio is just old
Minio is 1 second old
Minio is 9 seconds old
Minio is 2 hours old
Minio is 3 months old
Minio is 1 minute old
Minio is 253 years old
Minio is 1 second old
Minio is 1 minute old
```

There is one minor problem I saw is If a change in 
```go
	utcNow := time.Now().UTC()
	pastUTCNow := time.Now().UTC().Add(duration)
```
order of the line or `utcNow` is used to calculate `pastUTCNow`, accuracy is lost
```go
	utcNow := time.Now().UTC()
	pastUTCNow := utcNow.Add(duration)
```
or
```go
	pastUTCNow := time.Now().UTC().Add(duration)
	utcNow := time.Now().UTC()
```

result in duration = 1*time.Minute becomes 59 seconds as human readable form.

I feel `updateRelTimeMagnitude` is very cryptic to understand.

**Comment by harshavardhana** [comment]
> Nope. But you missed the point of update message. We found that there is an update available where current minio is just old. We would need to convey this message somehow for any non-zero duration.

I totally understand. If its about choosing the right message then `just` or `is` can be used as well since we can customize magnitudes however we feel like and to however precision.  

> I feel updateRelTimeMagnitude is very cryptic to understand.

We are using it from a library the library actually sufficiently documents this. In-fact one doesn't even need to use magnitudes you can just simply use `humanize.RelTime()` . The reason i am showing these examples is because we can change them to suite our current needs, including strings etc. 

> result in duration = 1*time.Minute becomes 59 seconds as human readable form.

You can optimize that further by choosing the right magnitudes if you need to optimize for that precision. One doesn't really care about this precision for an update message 1 minute old v/s 59 second doesn't make any difference. 

Current implementation exists because it was implemented when `humanize`  didn't have support for time. Now that `humanize` has and has implemented a much more richer functionality we can simply use that rather than maintaining one ourselves.

**Comment by balamurugana** [comment]
> I totally understand. If its about choosing the right message then just or is can be used as well since we can customize magnitudes however we feel like and to however precision.

Lets see what is decided for the UI

**Comment by nitisht** [comment]
@deekoder I understand this is pending for @abperiasamy review as there is a UX change.

**Comment by deekoder** [comment]
yes it is @NitishT - To be done today.

**Comment by deekoder** [comment]
@abperiasamy - Simply use https://github.com/dustin/go-humanize#times

```
fmt.Printf("You are running an older version of Minio released %s.", humanize.Time(buildTimeInstance))  
```

**Comment by harshavardhana** [comment]
> fmt.Printf("You are running an older version of Minio released %s.", humanize.Time(buildTimeInstance))

Since we have two times current build time and release time - we need to use `humanize.RelTime(latestReleaseTime,  currentReleaseTime, "ago", "from now")`

**Review by nitisht — APPROVED** [review]

**Comment by nitisht** [comment]
> Since we have two times current build time and release time - we need to use humanize.RelTime(latestReleaseTime, currentReleaseTime, "ago", "from now")

Can be reviewed once these changes are done.

**Comment by deekoder** [comment]
Can this be merged now @balamurugana ?

File: cmd/update-notifier.go [hunk 2]
```diff
@@ -36,14 +37,15 @@ func colorizeUpdateMessage(updateString string, newerThan time.Duration) string
 	// Calculate length without color coding, due to ANSI color
 	// characters padded to actual string the final length is wrong
 	// than the original string length.
-	hTime := timeDurationToHumanizedDuration(newerThan)
-	line1Str := fmt.Sprintf(" Minio is %s old ", hTime.StringShort())
+	newerThanStr := humanize.Time(UTCNow().Add(newerThan))
```
Comment (harshavardhana) [inline cmd/update-notifier.go:40]: We shouldn't use UTCNow() to calculate this, we should use the currentBuildTime v/s currentReleaseTime. using `humanize.RelTime()`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
> Can this be merged now @balamurugana ?

This cannot be merged yet see my review comment. @deekoder

File: cmd/update-notifier.go (hunk 2 above)
Comment (balamurugana) [reply cmd/update-notifier.go:40]: `currentBuildTime` and `currentReleaseTime` are in UTC.  `newerThan` is a difference of those two time.  In this function, we are printing how much newer `currentReleaseTime`.  Whether the duration is from `currentBuildTime` or `UTCNow()` makes no difference.

Comment (harshavardhana) [reply cmd/update-notifier.go:40]: Oh i see the time.Duration is used..

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
can you fix the CI builds @balamurugana ?

**Comment by harshavardhana** [comment]
One of the tests have failed @balamurugana 

```
--- FAIL: TestUpdateNotifier (0.00s)
	update-notifier_test.go:29: Duration string not found in colorized update message 
		┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
		┃ You are running an older version of Minio released 2 days from now ┃
		┃ Update: https://dl.minio.io/server/minio/release/linux-amd64/    ┃
		┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
		
```

**Comment by balamurugana** [comment]
The test passes locally but fails in Appveyor and travis CI.  Need to check why.

**Comment by harshavardhana** [comment]
> The test passes locally but fails in Appveyor and travis CI. Need to check why.

Probably the newer changes in humanize package are not vendorized locally? @balamurugana

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Unapproving until CI is fixed.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**